### PR TITLE
wip for legend refactor (part 2)

### PR DESCRIPTION
**Pull request by lindsayplatt**
- [x] New workflow in each graphics call 
- [x] Combine logic in set_legend_args - don't need to separate lines/pts as much
- [x] Combine polygon/rect logic 
- [x] Move legend.args into object>>legend (they should not be top-level) 
- [x] ~~Use `lazy_eval` instead of `quote`?~~ _more complicated than quote for this case_
- [x] ~~Add `lazy_eval` stuff from set_window_args to handle expressions~~ _doesn't need anything to handle expressions because it is only taking graphical parameters such as `lwd, lty, col` for the legend and ignores everything else._ 
- [x] Fix: using legend.name in error_bars gives a legend item for each error type (y.high, y.low, x.high, x.low)
- [x] ~~Fix `is.na(data.var)` warning from lines 141-151 of calc_views.R~~ _made into a separate issue since it is not caused by legend() - issue_ #256
- [x] Fix warning from having legend arguments in the `to.gsplot`variable in set_window_args
- [x] Fix/look into warning coming from base graphics
- [x] documentation for which args to use in legend vs graphics call
- [x] tests to check that overall legend arg static list (legend.R) & fun args static list (set_args.R) are what we expect

Weird thing with having error_bars use legend.name: for each part of the error bar (y.high, y.low, etc), an arrows call is made - therefore, a separate legend.name is added. This has two legend items:

``` r
gs <- gsplot()
gsNew <- points(gs, y=c(3,1,2), x=1:3, xlim=c(0,NA),ylim=c(0,NA),
                col="blue", pch=18, legend.name="Points")
gsNew <- lines(gsNew, c(3,4,3), c(2,4,6), legend.name="Lines")
gsNew <- abline(gsNew, b=1, a=0, legend.name="1:1")
gsNew <- legend(gsNew, location = "topleft",title="Awesome!")
gsNew <- grid(gsNew)
gsNew <- error_bar(gsNew, 1:3, y=c(3,1,2), y.high=c(0.5,0.25,1), y.low=0.1, legend.name="errors")
gsNew <- error_bar(gsNew, x=1:3, y=c(3,1,2), x.low=c(.2,NA,.2), x.high=.2, col="red",lwd=3)
gsNew <- title(gsNew, "Graphing Fun")
gsNew 
```

Warnings:

```
In is.na(data.var) :
  is.na() applied to non-(list or vector) of type 'NULL'

In (function (xleft, ybottom, xright, ytop, density = NULL, angle = 45,  :
  "text.font" is not a graphical parameter

In x1 + dx :
  longer object length is not a multiple of shorter object length
```

### CODE REVIEW

**Comment by lindsayplatt** [comment]
Transparent legend:

``` r
gsNew <- gsplot() %>% 
    rect(0,0,4,2, col="lightgreen", border=NA, legend.name="Shape") %>%
    points(y=c(3,1,2), x=1:3, xlim=c(0,NA),ylim=c(0,NA),
           col="blue", pch=18, legend.name="Points") %>% 
    lines(c(3,4,3), c(2,4,6), legend.name="Lines") %>% 
    abline(b=1, a=0, legend.name="1:1") %>% 
    legend(location = "topright", text.col="white", 
           box.lwd=2, ncol=2, bg=NA) %>% 
    grid() %>% 
    error_bar(1:3, y=c(3,1,2), y.high=c(0.5,0.25,1), y.low=0.1) %>% 
    error_bar(x=1:3, y=c(3,1,2), x.low=c(.2,NA,.2), x.high=.2, col="red",lwd=3) %>% 
    title("Graphing Fun")
gsNew
```

image

**Comment by lindsayplatt** [comment]
``` r
gs <- gsplot()  %>% 
    rect(1,5,2,7, col="pink", border=NA, legend.name="rectangle", text.font=3) %>%
    points(1:5, 1:5, pch=5, col="blue", legend.name="points") %>%
    lines(3:7, 4:8, col="darkgreen", legend.name="line") %>%
    error_bar(x=4, y=4, y.high=0.5, y.low=0.7, epsilon=0.5, legend.name="Errors") %>% 
    legend(x="bottomright", bg="grey", box.lty="dashed", box.lwd=3)
gs
```

image

**Comment by ldecicco-USGS** [comment]
<snifff > it's sooo pretty <\sniff >

I assume we only want 1 line for `error_bar`'s? 

Edit...I can't surround my text with humorous tags....github formats it out.

**Comment by lindsayplatt** [comment]
yep, I've got that in my list of things left to do!

**Comment by lindsayplatt** [comment]
Duplicated legend items for error bars is fixed:

``` r
gs <- gsplot()  %>% 
    rect(1,5,2,7, col="pink", border=NA, legend.name="rectangle", text.font=3) %>%
    points(1:5, 1:5, pch=5, col="blue", legend.name="points") %>%
    lines(3:7, 4:8, col="darkgreen", legend.name="line") %>%
    error_bar(x=4, y=4, y.high=0.5, y.low=0.7, epsilon=0.5, legend.name="Errors") %>% 
    legend(x="bottomright", bg="grey", box.lty="dashed", box.lwd=3)
gs
```

image

**Comment by lindsayplatt** [comment]
@ldecicco-USGS @jread-usgs @jiwalker-usgs @mhines-usgs 

Here is my rework of `legend` from the comments on Monday's code review. The new commits start at "new workflow for graphics calls". I have also included a couple of graphs that show off how easy it is to add legends in gsplot.

File: tests/testthat/test-legend.R [hunk 1]
```diff
@@ -23,6 +23,23 @@ test_that("gsplot legend works", {
 
   expect_equal(gs$legend$gs.config$location, "bottomright")
   
+})
+
+test_that("gsplot legend args are still the same", {
```
Comment (jiwalker-usgs) [inline tests/testthat/test-legend.R:34]: :+1:

**Comment by jordansread** [comment]
still a wip?

**Comment by lindsayplatt** [comment]
just waiting for a :+1: from reviewers I tagged :)

**Comment by jiwalker-usgs** [comment]
:+1:

**Comment by mhines-usgs** [comment]
I haven't pulled it down and played yet, but dang that's pretty! will grab and see what happens! :+1:

**Comment by jordansread** [comment]
:+1:

**Comment by lindsayplatt** [comment]
This should fix #244 #226 #219 #200 #160 #213

**Pull request merged**